File: src/completion_engine.rs
/*!
Universal completion engine for goobits-generated CLIs (Rust)
Reads goobits.yaml at runtime and provides context-aware completions
*/

use std::collections::HashMap;
use std::env;
use std::fs;
use std::path::Path;
use serde::{Deserialize, Serialize};
use serde_yaml;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CliOption {
    pub name: String,
    pub short: Option<String>,
    #[serde(rename = "type")]
    pub option_type: Option<String>,
    pub choices: Option<Vec<String>>,
    pub desc: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Argument {
    pub name: String,
    pub required: Option<bool>,
    pub choices: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Command {
    pub desc: Option<String>,
    pub args: Option<Vec<Argument>>,
    pub options: Option<Vec<CliOption>>,
    pub subcommands: Option<HashMap<String, Command>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CliConfig {
    pub commands: Option<HashMap<String, Command>>,
    pub options: Option<Vec<CliOption>>,
    pub enable_upgrade_command: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Config {
    pub cli: Option<CliConfig>,
}

#[derive(Debug, Clone)]
pub struct Context {
    pub context_type: ContextType,
    pub level: i32,
    pub current_command: Option<String>,
    pub subcommand: Option<String>,
    pub last_token: String,
    pub previous_token: String,
    pub partial: Option<String>,
    pub option: Option<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ContextType {
    Command,
    Subcommand,
    Option,
    OptionValue,
    Unknown,
}

pub struct CompletionEngine {
    config_path: String,
    config: Config,
}

impl CompletionEngine {
    pub fn new(config_path: Option<String>) -> Self {
        let config_path = config_path.unwrap_or_else(|| Self::find_config());
        let config = Self::load_config(&config_path);
        
        CompletionEngine {
            config_path,
            config,
        }
    }

    fn find_config() -> String {
        let mut current = env::current_dir().unwrap_or_else(|_| Path::new(".").to_path_buf());
        
        // Search up the directory tree
        loop {
            let config_file = current.join("goobits.yaml");
            if config_file.exists() {
                return config_file.to_string_lossy().to_string();
            }
            
            match current.parent() {
                Some(parent) => current = parent.to_path_buf(),
                None => break,
            }
        }
        
        // Fallback to current directory
        "./goobits.yaml".to_string()
    }

    fn load_config(config_path: &str) -> Config {
        match fs::read_to_string(config_path) {
            Ok(content) => {
                match serde_yaml::from_str(&content) {
                    Ok(config) => config,
                    Err(_) => Config { cli: None },
                }
            }
            Err(_) => Config { cli: None },
        }
    }

    pub fn get_completions(&self, _shell: &str, current_line: &str, cursor_pos: Option<usize>) -> Vec<String> {
        match self.get_completions_impl(current_line, cursor_pos) {
            Ok(completions) => {
                let mut unique_completions: Vec<String> = completions.into_iter().collect::<std::collections::HashSet<_>>().into_iter().collect();
                unique_completions.sort();
                unique_completions
            }
            Err(_) => Vec::new(),
        }
    }

    fn get_completions_impl(&self, current_line: &str, cursor_pos: Option<usize>) -> Result<Vec<String>, Box<dyn std::error::Error>> {
        // Parse the command line
        let tokens = self.parse_command_line(current_line, cursor_pos)?;
        let context = self.analyze_context(&tokens);
        
        // Generate completions based on context
        self.generate_completions(&context)
    }

    fn parse_command_line(&self, line: &str, cursor_pos: Option<usize>) -> Result<Vec<String>, Box<dyn std::error::Error>> {
        let cursor_pos = cursor_pos.unwrap_or(line.len());
        
        // Extract the relevant part of the line up to cursor
        let relevant_line = &line[..cursor_pos.min(line.len())];
        
        // Simple tokenization (could be enhanced for complex cases)
        let mut tokens = Vec::new();
        let mut current_token = String::new();
        let mut in_quotes = false;
        let mut quote_char = None;
        
        for ch in relevant_line.chars() {
            match ch {
                '"' | '\'' if !in_quotes => {
                    in_quotes = true;
                    quote_char = Some(ch);
                    current_token.push(ch);
                }
                ch if Some(ch) == quote_char && in_quotes => {
                    in_quotes = false;
                    current_token.push(ch);
                    quote_char = None;
                }
                ch if ch.is_whitespace() && !in_quotes => {
                    if !current_token.is_empty() {
                        tokens.push(current_token.clone());
                        current_token.clear();
                    }
                }
                ch => {
                    current_token.push(ch);
                }
            }
        }
        
        // Add final token if exists
        if !current_token.is_empty() {
            tokens.push(current_token);
        }
        
        Ok(tokens)
    }

    fn analyze_context(&self, tokens: &[String]) -> Context {
        if tokens.is_empty() {
            return Context {
                context_type: ContextType::Command,
                level: 0,
                current_command: None,
                subcommand: None,
                last_token: String::new(),
                previous_token: String::new(),
                partial: None,
                option: None,
            };
        }
        
        // Skip the program name
        let tokens = if tokens.len() > 0 { &tokens[1..] } else { tokens };
        
        if tokens.is_empty() {
            return Context {
                context_type: ContextType::Command,
                level: 0,
                current_command: None,
                subcommand: None,
                last_token: String::new(),
                previous_token: String::new(),
                partial: None,
                option: None,
            };
        }
        
        let last_token = tokens.last().cloned().unwrap_or_default();
        let previous_token = if tokens.len() > 1 {
            tokens[tokens.len() - 2].clone()
        } else {
            String::new()
        };
        
        let mut context = Context {
            context_type: ContextType::Unknown,
            level: 0,
            current_command: None,
            subcommand: None,
            last_token: last_token.clone(),
            previous_token: previous_token.clone(),
            partial: None,
            option: None,
        };
        
        // Check if we're completing an option
        if last_token.starts_with('-') {
            context.context_type = ContextType::Option;
            return context;
        }
        
        // Check if previous token was an option that expects a value
        if previous_token.starts_with('-') {
            context.context_type = ContextType::OptionValue;
            context.option = Some(previous_token);
            return context;
        }
        
        // Determine command/subcommand context
        let empty_commands = HashMap::new();
        let cli_commands = match &self.config.cli {
            Some(cli) => cli.commands.as_ref().unwrap_or(&empty_commands),
            None => &empty_commands,
        };
        
        if tokens.len() == 1 {
            // Completing top-level command
            context.context_type = ContextType::Command;
            context.level = 0;
            context.partial = Some(tokens[0].clone());
        } else if let Some(command_config) = cli_commands.get(&tokens[0]) {
            // We have a valid command, check for subcommands
            context.current_command = Some(tokens[0].clone());
            
            if command_config.subcommands.is_some() && tokens.len() == 2 {
                context.context_type = ContextType::Subcommand;
                context.level = 1;
                context.partial = Some(tokens[1].clone());
            } else if let Some(subcommands) = &command_config.subcommands {
                if tokens.len() > 2 && subcommands.contains_key(&tokens[1]) {
                    context.context_type = ContextType::Option;
                    context.subcommand = Some(tokens[1].clone());
                    context.level = 2;
                } else {
                    context.context_type = ContextType::Option;
                    context.level = 1;
                }
            } else {
                context.context_type = ContextType::Option;
                context.level = 1;
            }
        }
        
        context
    }

    fn generate_completions(&self, context: &Context) -> Result<Vec<String>, Box<dyn std::error::Error>> {
        let mut completions = Vec::new();
        
        match context.context_type {
            ContextType::Command => {
                completions.extend(self.get_command_completions(context));
            }
            ContextType::Subcommand => {
                completions.extend(self.get_subcommand_completions(context));
            }
            ContextType::Option => {
                completions.extend(self.get_option_completions(context));
            }
            ContextType::OptionValue => {
                completions.extend(self.get_option_value_completions(context));
            }
            ContextType::Unknown => {}
        }
        
        // Filter by partial match if applicable
        if let Some(partial) = &context.partial {
            let filtered: Vec<String> = completions
                .into_iter()
                .filter(|c| c.starts_with(partial))
                .collect();
            return Ok(filtered);
        }
        
        Ok(completions)
    }

    fn get_command_completions(&self, _context: &Context) -> Vec<String> {
        let mut commands = Vec::new();
        
        // CLI commands from config
        if let Some(cli) = &self.config.cli {
            if let Some(cli_commands) = &cli.commands {
                commands.extend(cli_commands.keys().cloned());
            }
        }
        
        // Built-in commands
        let mut builtin_commands = vec!["help", "version", "completion", "config"];
        
        // Check if upgrade is enabled
        let upgrade_enabled = self.config.cli
            .as_ref()
            .and_then(|cli| cli.enable_upgrade_command)
            .unwrap_or(true);
        
        if upgrade_enabled {
            builtin_commands.push("upgrade");
        }
        
        commands.extend(builtin_commands.into_iter().map(String::from));
        
        commands
    }

    fn get_subcommand_completions(&self, context: &Context) -> Vec<String> {
        let command = match &context.current_command {
            Some(cmd) => cmd,
            None => return Vec::new(),
        };
        
        // Handle built-in commands first
        match command.as_str() {
            "completion" => {
                return vec![
                    "generate".to_string(),
                    "install".to_string(),
                    "instructions".to_string(),
                ];
            }
            "config" => {
                return vec![
                    "get".to_string(),
                    "set".to_string(),
                    "reset".to_string(),
                    "path".to_string(),
                ];
            }
            _ => {}
        }
        
        // Handle user-defined commands
        let empty_commands = HashMap::new();
        let cli_commands = match &self.config.cli {
            Some(cli) => cli.commands.as_ref().unwrap_or(&empty_commands),
            None => return Vec::new(),
        };
        
        let command_config = match cli_commands.get(command) {
            Some(config) => config,
            None => return Vec::new(),
        };
        
        let subcommands = match &command_config.subcommands {
            Some(subs) => subs,
            None => return Vec::new(),
        };
        
        subcommands.keys().cloned().collect()
    }

    fn get_option_completions(&self, context: &Context) -> Vec<String> {
        let mut options = Vec::new();
        
        // Global options
        if let Some(cli) = &self.config.cli {
            if let Some(global_options) = &cli.options {
                for opt in global_options {
                    options.push(format!("--{}", opt.name));
                    if let Some(short) = &opt.short {
                        options.push(format!("-{}", short));
                    }
                }
            }
        }
        
        // Command-specific options
        if let Some(command) = &context.current_command {
            let empty_commands = HashMap::new();
            let cli_commands = match &self.config.cli {
                Some(cli) => cli.commands.as_ref().unwrap_or(&empty_commands),
                None => &empty_commands,
            };
            
            if let Some(mut command_config) = cli_commands.get(command).cloned() {
                // Check if we're in a subcommand
                if let Some(subcommand) = &context.subcommand {
                    if let Some(subcommands) = &command_config.subcommands {
                        if let Some(sub_config) = subcommands.get(subcommand) {
                            command_config = sub_config.clone();
                        }
                    }
                }
                
                if let Some(command_options) = &command_config.options {
                    for opt in command_options {
                        options.push(format!("--{}", opt.name));
                        if let Some(short) = &opt.short {
                            options.push(format!("-{}", short));
                        }
                    }
                }
            }
        }
        
        // Standard options
        options.extend(vec![
            "--help".to_string(),
            "-h".to_string(),
            "--version".to_string(),
            "-V".to_string(),
        ]);
        
        options
    }

    fn get_option_value_completions(&self, context: &Context) -> Vec<String> {
        let option = match &context.option {
            Some(opt) => opt.trim_start_matches('-'),
            None => return Vec::new(),
        };
        
        // Handle built-in command option completions
        if let Some(command) = &context.current_command {
            match (command.as_str(), option) {
                ("completion", "shell") => {
                    return vec![
                        "bash".to_string(),
                        "zsh".to_string(),
                        "fish".to_string(),
                        "powershell".to_string(),
                    ];
                }
                ("config", "format") => {
                    return vec![
                        "json".to_string(),
                        "yaml".to_string(),
                        "plain".to_string(),
                    ];
                }
                ("config", "value-type") => {
                    return vec![
                        "string".to_string(),
                        "int".to_string(),
                        "float".to_string(),
                        "bool".to_string(),
                    ];
                }
                _ => {}
            }
        }
        
        // Find option configuration for user-defined commands
        let option_config = match self.find_option_config(option, context) {
            Some(config) => config,
            None => return Vec::new(),
        };
        
        let option_type = option_config.option_type.as_deref().unwrap_or("str");
        
        // Handle different option types
        if option_type == "choice" {
            if let Some(choices) = &option_config.choices {
                return choices.clone();
            }
        } else if option_type == "file" || matches!(option, "config" | "file" | "input" | "output") {
            return self.get_file_completions();
        } else if option_type == "dir" || matches!(option, "directory" | "dir" | "output-dir") {
            return self.get_directory_completions();
        } else if option_type == "bool" || option_type == "flag" {
            return vec!["true".to_string(), "false".to_string()];
        }
        
        Vec::new()
    }

    fn find_option_config(&self, option_name: &str, context: &Context) -> Option<CliOption> {
        // Check global options
        if let Some(cli) = &self.config.cli {
            if let Some(global_options) = &cli.options {
                for opt in global_options {
                    if opt.name == option_name || opt.short.as_deref() == Some(option_name) {
                        return Some(opt.clone());
                    }
                }
            }
        }
        
        // Check command-specific options
        if let Some(command) = &context.current_command {
            let empty_commands = HashMap::new();
            let cli_commands = match &self.config.cli {
                Some(cli) => cli.commands.as_ref().unwrap_or(&empty_commands),
                None => return None,
            };
            
            if let Some(mut command_config) = cli_commands.get(command).cloned() {
                // Check if we're in a subcommand
                if let Some(subcommand) = &context.subcommand {
                    if let Some(subcommands) = &command_config.subcommands {
                        if let Some(sub_config) = subcommands.get(subcommand) {
                            command_config = sub_config.clone();
                        }
                    }
                }
                
                if let Some(command_options) = &command_config.options {
                    for opt in command_options {
                        if opt.name == option_name || opt.short.as_deref() == Some(option_name) {
                            return Some(opt.clone());
                        }
                    }
                }
            }
        }
        
        None
    }

    fn get_file_completions(&self) -> Vec<String> {
        match fs::read_dir(".") {
            Ok(entries) => {
                let mut files = Vec::new();
                for entry in entries.flatten() {
                    if let Ok(metadata) = entry.metadata() {
                        if metadata.is_file() {
                            if let Some(name) = entry.file_name().to_str() {
                                files.push(name.to_string());
                            }
                        }
                    }
                }
                files
            }
            Err(_) => Vec::new(),
        }
    }

    fn get_directory_completions(&self) -> Vec<String> {
        match fs::read_dir(".") {
            Ok(entries) => {
                let mut dirs = Vec::new();
                for entry in entries.flatten() {
                    if let Ok(metadata) = entry.metadata() {
                        if metadata.is_dir() {
                            if let Some(name) = entry.file_name().to_str() {
                                dirs.push(format!("{}/", name));
                            }
                        }
                    }
                }
                dirs
            }
            Err(_) => Vec::new(),
        }
    }
}

// Main completion function - called by CLIs with _completion command
pub fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 4 {
        std::process::exit(1);
    }
    
    let shell = &args[2];
    let current_line = &args[3];
    let cursor_pos = if args.len() > 4 {
        args[4].parse().ok()
    } else {
        None
    };
    
    let engine = CompletionEngine::new(None);
    let completions = engine.get_completions(shell, current_line, cursor_pos);
    
    // Output completions one per line
    for completion in completions {
        println!("{}", completion);
    }
}

File: src/goobits_cli/completion_engine.py
#!/usr/bin/env python3
"""
Universal completion engine for goobits-generated CLIs
Reads goobits.yaml at runtime and provides context-aware completions
"""

import os
import sys
import yaml
import json
from pathlib import Path
from typing import Dict, List, Optional, Any, Tuple


class CompletionEngine:
    """Universal completion engine that parses goobits.yaml at runtime"""
    
    def __init__(self, config_path: Optional[str] = None):
        """Initialize completion engine with goobits.yaml path"""
        self.config_path = config_path or self._find_config()
        self.config = self._load_config()
        
    def _find_config(self) -> str:
        """Find goobits.yaml file in current directory or parent directories"""
        current = Path.cwd()
        
        # Search up the directory tree
        for parent in [current] + list(current.parents):
            config_file = parent / "goobits.yaml"
            if config_file.exists():
                return str(config_file)
                
        # Fallback to current directory
        return str(Path.cwd() / "goobits.yaml")
    
    def _load_config(self) -> Dict[str, Any]:
        """Load and parse goobits.yaml configuration"""
        try:
            with open(self.config_path, 'r', encoding='utf-8') as f:
                return yaml.safe_load(f) or {}
        except Exception:
            return {}
    
    def get_completions(self, shell: str, current_line: str, cursor_pos: Optional[int] = None) -> List[str]:
        """Get completions for the current shell context"""
        try:
            # Parse the command line
            tokens = self._parse_command_line(current_line, cursor_pos)
            
            # Special case: if the line ends with a space after an option, 
            # we should be completing the option value
            if current_line.endswith(' ') and len(tokens) > 0 and tokens[-1].startswith('-'):
                # Add empty token to represent the word being completed
                tokens.append('')
            
            context = self._analyze_context(tokens)
            
            # Generate completions based on context
            completions = self._generate_completions(context)
            
            return sorted(set(completions))
        except Exception:
            return []
    
    def _parse_command_line(self, line: str, cursor_pos: Optional[int] = None) -> List[str]:
        """Parse command line into tokens, handling quoted arguments"""
        if cursor_pos is None:
            cursor_pos = len(line)
            
        # Extract the relevant part of the line up to cursor
        relevant_line = line[:cursor_pos]
        
        # Simple tokenization (could be enhanced for complex cases)
        tokens = []
        current_token = ""
        in_quotes = False
        quote_char = None
        
        for char in relevant_line:
            if char in ['"', "'"] and not in_quotes:
                in_quotes = True
                quote_char = char
                current_token += char
            elif char == quote_char and in_quotes:
                in_quotes = False
                current_token += char
                quote_char = None
            elif char.isspace() and not in_quotes:
                if current_token:
                    tokens.append(current_token)
                    current_token = ""
            else:
                current_token += char
        
        # Add final token if exists
        if current_token:
            tokens.append(current_token)
            
        return tokens
    
    def _analyze_context(self, tokens: List[str]) -> Dict[str, Any]:
        """Analyze the current completion context"""
        if not tokens:
            return {"type": "command", "level": 0}
            
        # Skip the program name
        if len(tokens) > 0:
            tokens = tokens[1:]
            
        if not tokens:
            return {"type": "command", "level": 0}
            
        context = {
            "type": "unknown",
            "level": 0,
            "current_command": None,
            "subcommand": None,
            "last_token": tokens[-1] if tokens else "",
            "previous_token": tokens[-2] if len(tokens) > 1 else "",
        }
        
        # Determine command/subcommand context first
        cli_commands = self.config.get("cli", {}).get("commands", {})
        
        # Find the current command in the token list
        for i, token in enumerate(tokens):
            if token in cli_commands:
                context["current_command"] = token
                break
        
        # Check if we're completing an option
        if context["last_token"].startswith("-"):
            context["type"] = "option"
            return context
            
        # Check if previous token was an option that expects a value
        if context["previous_token"].startswith("-") and not context["last_token"].startswith("-"):
            context["type"] = "option_value"
            context["option"] = context["previous_token"]
            return context
        
        if len(tokens) == 1:
            # Completing top-level command
            context["type"] = "command"
            context["level"] = 0
            context["partial"] = tokens[0]
        elif len(tokens) >= 2 and tokens[0] in cli_commands:
            # We have a valid command, check for subcommands
            command_config = cli_commands[tokens[0]]
            context["current_command"] = tokens[0]
            
            if "subcommands" in command_config and len(tokens) == 2 and not tokens[1].startswith("-"):
                context["type"] = "subcommand"
                context["level"] = 1
                context["partial"] = tokens[1]
            elif "subcommands" in command_config and len(tokens) > 2 and tokens[1] in command_config["subcommands"]:
                context["type"] = "option" if context["type"] == "unknown" else context["type"]
                context["subcommand"] = tokens[1]
                context["level"] = 2
            else:
                # We're in option context for this command
                context["type"] = "option" if context["type"] == "unknown" else context["type"]
                context["level"] = 1
        elif len(tokens) >= 1:
            # First token is not a recognized command, treat as partial command
            context["type"] = "command"
            context["level"] = 0
            context["partial"] = tokens[0]
        
        return context
    
    def _generate_completions(self, context: Dict[str, Any]) -> List[str]:
        """Generate completions based on context"""
        completions = []
        
        if context["type"] == "command":
            completions.extend(self._get_command_completions(context))
        elif context["type"] == "subcommand":
            completions.extend(self._get_subcommand_completions(context))
        elif context["type"] == "option":
            completions.extend(self._get_option_completions(context))
        elif context["type"] == "option_value":
            completions.extend(self._get_option_value_completions(context))
            
        # Filter by partial match if applicable
        partial = context.get("partial", "")
        if partial:
            completions = [c for c in completions if c.startswith(partial)]
            
        return completions
    
    def _get_command_completions(self, context: Dict[str, Any]) -> List[str]:
        """Get top-level command completions"""
        commands = []
        
        # CLI commands from config
        cli_commands = self.config.get("cli", {}).get("commands", {})
        commands.extend(cli_commands.keys())
        
        # Built-in commands
        builtin_commands = ["help", "version"]
        
        # Check if upgrade is enabled
        cli_config = self.config.get("cli", {})
        if cli_config.get("enable_upgrade_command", True):
            builtin_commands.append("upgrade")
            
        commands.extend(builtin_commands)
        
        return commands
    
    def _get_subcommand_completions(self, context: Dict[str, Any]) -> List[str]:
        """Get subcommand completions"""
        command = context.get("current_command")
        if not command:
            return []
            
        cli_commands = self.config.get("cli", {}).get("commands", {})
        command_config = cli_commands.get(command, {})
        
        subcommands = command_config.get("subcommands", {})
        return list(subcommands.keys())
    
    def _get_option_completions(self, context: Dict[str, Any]) -> List[str]:
        """Get option completions for current command/subcommand"""
        options = []
        
        # Global options
        global_options = self.config.get("cli", {}).get("options", [])
        for opt in global_options:
            options.append(f"--{opt['name']}")
            if opt.get("short"):
                options.append(f"-{opt['short']}")
        
        # Command-specific options
        command = context.get("current_command")
        if command:
            cli_commands = self.config.get("cli", {}).get("commands", {})
            command_config = cli_commands.get(command, {})
            
            # Check if we're in a subcommand
            subcommand = context.get("subcommand")
            if subcommand and "subcommands" in command_config:
                command_config = command_config["subcommands"].get(subcommand, {})
            
            command_options = command_config.get("options", [])
            for opt in command_options:
                options.append(f"--{opt['name']}")
                if opt.get("short"):
                    options.append(f"-{opt['short']}")
        
        # Standard options
        options.extend(["--help", "-h", "--version", "-V"])
        
        return options
    
    def _get_option_value_completions(self, context: Dict[str, Any]) -> List[str]:
        """Get completions for option values"""
        option = context.get("option", "").lstrip("-")
        
        # Find option configuration
        option_config = self._find_option_config(option, context)
        if not option_config:
            return []
        
        option_type = option_config.get("type", "str")
        
        # Handle different option types
        if option_type == "choice" and "choices" in option_config:
            return option_config["choices"]
        elif option_type == "file" or option in ["config", "file", "input"]:
            return self._get_file_completions()
        elif option_type == "dir" or option in ["directory", "dir", "output-dir"]:
            return self._get_directory_completions()
        elif option_type == "bool" or option_type == "flag":
            return ["true", "false"]
            
        return []
    
    def _find_option_config(self, option_name: str, context: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Find configuration for a specific option"""
        # Check global options
        global_options = self.config.get("cli", {}).get("options", [])
        for opt in global_options:
            if opt["name"] == option_name or opt.get("short") == option_name:
                return opt
        
        # Check command-specific options
        command = context.get("current_command")
        if command:
            cli_commands = self.config.get("cli", {}).get("commands", {})
            command_config = cli_commands.get(command, {})
            
            # Check if we're in a subcommand
            subcommand = context.get("subcommand")
            if subcommand and "subcommands" in command_config:
                command_config = command_config["subcommands"].get(subcommand, {})
            
            command_options = command_config.get("options", [])
            for opt in command_options:
                if opt["name"] == option_name or opt.get("short") == option_name:
                    return opt
        
        return None
    
    def _get_file_completions(self) -> List[str]:
        """Get file completions for current directory"""
        try:
            files = []
            for item in os.listdir("."):
                if os.path.isfile(item):
                    files.append(item)
            return files
        except Exception:
            return []
    
    def _get_directory_completions(self) -> List[str]:
        """Get directory completions for current directory"""
        try:
            dirs = []
            for item in os.listdir("."):
                if os.path.isdir(item):
                    dirs.append(item + "/")
            return dirs
        except Exception:
            return []


def main():
    """Main completion function - called by CLIs with --completion"""
    if len(sys.argv) < 3:
        sys.exit(1)
        
    shell = sys.argv[1]
    current_line = sys.argv[2]
    cursor_pos = int(sys.argv[3]) if len(sys.argv) > 3 else None
    
    engine = CompletionEngine()
    completions = engine.get_completions(shell, current_line, cursor_pos)
    
    # Output completions one per line
    for completion in completions:
        print(completion)


if __name__ == "__main__":
    main()